Avoiding translation calls for the default locale

Vitalij-Voronkoff opened the pull request:
BroadleafCommerce/QA#3782

Vitalij-Voronkoff committed:
avoiding translation calls for the default locale

ckittrell commented on common/src/main/java/org/broadleafcommerce/common/i18n/service/DynamicTranslationProvider.java:
```diff
@@ -32,18 +32,22 @@
     /**
      * If translations are enabled, this method will look for a translation for the specified field. If translations are
      * disabled or if this particular field did not have a translation, it will return back the defaultValue.
+     * If the passed in locale's language matches the Broadleaf default locale, it will return back the defaultValue.
```
```suggestion
     * If in the context of the default locale, then skip all translation calls and return the defaultValue.
```

ckittrell reviewed (approved):
Approving with the expectation that we cleanup the javadoc message

ckittrell commented on common/src/main/java/org/broadleafcommerce/common/i18n/service/DynamicTranslationProvider.java:
```diff
@@ -32,18 +32,22 @@
+
+        // avoiding translation calls for the default locale
+        if (TranslationConsiderationContext.hasTranslation() &&
+                !requestContext.getLocale().getDefaultFlag()) {
```
@Vitalij-Voronkoff On second thought, I think we should refactor this to be `&& localeShouldBeTranslated(requestContext.getLocale())`

Within that method, we can simply return `!locale.getDefaultFlag()`, but it would give users the ability to declare that `en_us` & `en_gb` should not be translated if only `en_us` is the default.

ckittrell replied:
... realizing that `getValue(...)` is a static method which really derails the idea of extensibility

ckittrell committed:
- Update DynamicTranslationProvider javadocs
- Introduce method to TranslationService to declare whether or not the provided locale should be translated. By default, the default locale is the only one that should not be translated.

BroadleafCommerce/QA#3782

ckittrell commented:
@danielcolgrove @jfleschler I made a fair amount of changes on this one, so I'd like a second pair of eyes on it from one of you guys.

jfleschler commented on common/src/main/java/org/broadleafcommerce/common/i18n/service/TranslationServiceImpl.java:
```diff
@@ -492,7 +510,7 @@ protected boolean propertyInDefaultLocaleExceptionList(Object entity, String pro
      * @return
      */
-    protected boolean localeMatchesDefaultLocale(Locale locale) {
+    protected boolean languageMatchesDefaultLocaleLanguage(Locale locale) {
```
This is an API change that could break clients.  We probably don't want to have this as part of the PR

ckittrell committed:
- Revert name change that could break client code

BroadleafCommerce/QA#3782

ckittrell committed:
- Update TranslationServiceTest to include the creation of locales
- Throw an IllegalArgumentException in shouldTranslateLocale if the locale cannot be found for the given code

BroadleafCommerce/QA#3782

jfleschler reviewed (approved):
LGTM
